k9mail/src/main/java/com/fsck/k9/controller/QresyncSyncInteractor.java

```diff
@@ -0,0 +1,128 @@
+        final AtomicInteger headerProgress = new AtomicInteger(0);
+
+        for (MessagingListener l : controller.getListeners(listener)) {
+            l.synchronizeMailboxHeadersStarted(account, folderName);
+        }
+
+        List<ImapMessage> modifiedMessages = qresyncParamResponse.getModifiedMessages();
+        Timber.v("SYNC: Received %d FETCH responses in QRESYNC response", modifiedMessages.size());
+
+        String uid = localFolder.getSmallestMessageUid();
+        long smallestLocalUid = 1;
+        if (uid != null) {
+            smallestLocalUid = Long.parseLong(uid);
+        }
+
+        for (ImapMessage imapMessage : modifiedMessages) {
+            long remoteMessageUid = Long.parseLong(imapMessage.getUid());
+            Message localMessage = localFolder.getMessage(imapMessage.getUid());
+
+            if (remoteMessageUid < smallestLocalUid) {
+                continue;
+            }
+
+            if (localMessage == null || (!localMessage.isSet(Flag.X_DOWNLOADED_FULL) &&
+                    !localMessage.isSet(Flag.X_DOWNLOADED_PARTIAL))) {
+                remoteMessages.add(imapMessage);
+            } else {
+                messageDownloader.processDownloadedFlags(account, localFolder, imapMessage, headerProgress,
+                        modifiedMessages.size());
+            }
+        }
+
+        Timber.v("SYNC: Received %d new message UIDs in QRESYNC response", remoteMessages.size());
+    }
+
+    private void findOldRemoteMessagesToDownload(List<ImapMessage> remoteMessages) throws MessagingException {
+        /*At ths point, UIDs and flags for new messages and existing changed messages have been fetched, so all N
+          messages in the local store should correspond to the most recent N messages in the remote store. We still need
+          to download historical messages (messages older than the oldest message in the inbox) if necessary till we
+          reach the mailbox capacity.*/
+
+        String folderName = imapFolder.getName();
+        Date earliestDate = account.getEarliestPollDate();
+        final AtomicInteger headerProgress = new AtomicInteger(0);
+
+        int oldMessagesStart = ImapSyncInteractor.getRemoteStart(localFolder, imapFolder);
+        int oldMessagesEnd = oldMessagesStart - 1 + localFolder.getVisibleLimit() - remoteMessages.size() -
+                localFolder.getMessageCount();
+
+        if (oldMessagesEnd > oldMessagesStart) {
+            List<ImapMessage> oldMessages = imapFolder.getMessages(oldMessagesStart, oldMessagesEnd, earliestDate, null);
+            Timber.v("SYNC: Fetched %d old message UIDs for folder %s", oldMessages.size(), folderName);
+            for (ImapMessage oldMessage : oldMessages) {
+                headerProgress.incrementAndGet();
+                for (MessagingListener l : controller.getListeners(listener)) {
+                    l.synchronizeMailboxHeadersProgress(account, folderName, headerProgress.get(), oldMessages.size());
```

Comment:
It does, through the `MessageDownloader.processDownloadedFlags` method. I've move the notification lines back to their methods to make it easier to understand.